paywall/src/data-iframe/blockchainHandler/keyStatus.js

```diff
@@ -0,0 +1,88 @@
+import { getAccount } from './account'
+
+/**
+ * Determine whether a key is valid based solely on its expiration
+ *
+ * @oaram {key} key a key object, which must contain an expiration timestamp in seconds
+ * @returns bool
+ */
+export function validKey(key) {
+  return key.expiration > new Date().getTime() / 1000
+}
+
+/**
+ * Returns the textual status of a key based on its expiration and transaction
+ * @param {key} key a key object, which must contain transactions and expiration
+ * @param {int} requiredConfirmations the number of confirmations required to consider a transaction confirmed
+ * @returns {none|submitted|pending|confirming|valid|expired|failed}
+ */
+export function getKeyStatus(key, requiredConfirmations) {
+  const valid = validKey(key)
+  if (!key.transactions || !key.transactions.length) {
+    return valid ? 'valid' : 'none'
+  }
+  // if we reach this line, key.transactions has at least 1 transaction in it
+  // the key transactions must be generated by linkTransactionsToKeys in order
+  // to be properly sorted
+  const lastTransaction = key.transactions[0]
+  switch (lastTransaction.status) {
+    case 'mined':
+      if (lastTransaction.confirmations < requiredConfirmations) {
+        return valid ? 'confirming' : 'expired'
+      }
+      return valid ? 'valid' : 'expired'
+    default:
+      // NOTE: a key may technically expire while the transaction is pending, but
+      // its timer only actually starts when the transaction is first confirmed
+      // so users get the pending time for free
+      return lastTransaction.status || 'none'
+  }
+}
+
+/**
+ * Construct the transactions field for each key
+ * @param {object} keys keys, indexed by their lock/owner ID (not the same as smart contract ID)
+ * @param {object} transactions transactions, indexed by hash (submitted transaction is always indexed under "submitted")
```

Comment:
Not sure I understand that? submitted transactions should have a hash too?

Reply:
There are 3 stages of transaction that we have to manage:

1. submitted - the transaction doesn't exist in the mempool yet, has no hash or block number
2. pending - the transaction exists in the mempool, has a hash, and no block number
3. mined - has a hash and a block number

The first state means we have to index by something to put it in transactions, so I decided (and this is a change from the original work) to index by ``submitted-${lock}-${account}`` so that we can have multiple submitted transactions if the user is weird and does this in many open tabs all at the same time for different locks.

Once the transaction is in the mempool (which should be very fast, but occasionally alchemy doesn't), we have a hash and the submitted placeholder can be removed